File: go/pkg/identity/identity.go
// Package identity models the trusted identity context that the Gateway injects
// and that downstream services consume.
//
// Rules (see docs/AI_CONTEXT.md):
//   - The Gateway is the only component that may set these headers from verified
//     tokens. It MUST strip any client-supplied values first (StripUntrusted).
//   - Business services NEVER parse end-user JWTs; they read the identity from
//     these headers (FromHeaders) and trust them only on internal traffic.
package identity

import (
	"context"
	"net/http"
	"strings"
)

// Trusted identity headers. Keep in sync with
// platform-contracts/schemas/identity-context.schema.json.
const (
	HeaderRequestID   = "X-Request-Id"
	HeaderUserID      = "X-User-Id"
	HeaderTenantID    = "X-Tenant-Id"
	HeaderRoles       = "X-Roles"       // comma-separated
	HeaderScopes      = "X-Scopes"      // comma-separated
	HeaderAuthSubject = "X-Auth-Subject"
)

// allHeaders is the full set the Gateway strips from external requests.
var allHeaders = []string{
	HeaderUserID, HeaderTenantID, HeaderRoles, HeaderScopes, HeaderAuthSubject, HeaderRequestID,
}

// Identity is the actor context carried with every request and internal call.
type Identity struct {
	RequestID string
	UserID    string
	TenantID  string
	Roles     []string
	Scopes    []string
	Subject   string
}

// StripUntrusted removes any client-supplied identity headers. The Gateway calls
// this on inbound external requests before injecting verified values.
func StripUntrusted(h http.Header) {
	for _, k := range allHeaders {
		h.Del(k)
	}
}

// Inject writes the identity onto outgoing headers (Gateway -> service, or
// service -> service propagation).
func (id Identity) Inject(h http.Header) {
	setNonEmpty(h, HeaderRequestID, id.RequestID)
	setNonEmpty(h, HeaderUserID, id.UserID)
	setNonEmpty(h, HeaderTenantID, id.TenantID)
	setNonEmpty(h, HeaderRoles, strings.Join(id.Roles, ","))
	setNonEmpty(h, HeaderScopes, strings.Join(id.Scopes, ","))
	setNonEmpty(h, HeaderAuthSubject, id.Subject)
}

// FromHeaders reads the identity from trusted (internal) request headers.
func FromHeaders(h http.Header) Identity {
	return Identity{
		RequestID: h.Get(HeaderRequestID),
		UserID:    h.Get(HeaderUserID),
		TenantID:  h.Get(HeaderTenantID),
		Roles:     splitCSV(h.Get(HeaderRoles)),
		Scopes:    splitCSV(h.Get(HeaderScopes)),
		Subject:   h.Get(HeaderAuthSubject),
	}
}

type ctxKey struct{}

// NewContext stores the identity in the context.
func NewContext(ctx context.Context, id Identity) context.Context {
	return context.WithValue(ctx, ctxKey{}, id)
}

// FromContext retrieves the identity from the context.
func FromContext(ctx context.Context) (Identity, bool) {
	id, ok := ctx.Value(ctxKey{}).(Identity)
	return id, ok
}

// Middleware extracts the identity from trusted headers into the context. Use it
// in business services, which must only see internal Gateway/service traffic.
func Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		ctx := NewContext(r.Context(), FromHeaders(r.Header))
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}

func setNonEmpty(h http.Header, k, v string) {
	if v != "" {
		h.Set(k, v)
	}
}

func splitCSV(s string) []string {
	if s == "" {
		return nil
	}
	parts := strings.Split(s, ",")
	out := parts[:0]
	for _, p := range parts {
		if p = strings.TrimSpace(p); p != "" {
			out = append(out, p)
		}
	}
	return out
}

File: node/apps/business-service/src/common/identity/identity.ts
export const IDENTITY_HEADERS = {
  requestId: 'x-request-id',
  userId: 'x-user-id',
  tenantId: 'x-tenant-id',
  roles: 'x-roles',
  scopes: 'x-scopes',
  subject: 'x-auth-subject',
} as const;

export type Identity = {
  requestId: string;
  userId: string;
  tenantId: string;
  roles: string[];
  scopes: string[];
  subject: string;
};

export function splitCsv(value: string | undefined): string[] {
  if (!value) {
    return [];
  }
  return value
    .split(',')
    .map((s) => s.trim())
    .filter(Boolean);
}

export function identityFromHeaders(headers: Record<string, string | string[] | undefined>): Identity {
  const pick = (key: string): string => {
    const raw = headers[key];
    if (Array.isArray(raw)) {
      return raw[0] ?? '';
    }
    return raw ?? '';
  };

  return {
    requestId: pick(IDENTITY_HEADERS.requestId),
    userId: pick(IDENTITY_HEADERS.userId),
    tenantId: pick(IDENTITY_HEADERS.tenantId),
    roles: splitCsv(pick(IDENTITY_HEADERS.roles)),
    scopes: splitCsv(pick(IDENTITY_HEADERS.scopes)),
    subject: pick(IDENTITY_HEADERS.subject),
  };
}

export function isAuthenticated(id: Identity | undefined): boolean {
  return Boolean(id?.userId);
}

export function hasRole(id: Identity | undefined, role: string): boolean {
  return id?.roles.includes(role) ?? false;
}
